wpilib/src/relay.rs

```diff
@@ -0,0 +1,216 @@
+// Copyright 2019 First Rust Competition Developers.
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Interface to the relay pins.
+//!
+//! Relays are intended to be connected to Spikes or similar relays.
+//! The relay channels control a pair of pins which can be independently
+//! toggled on or off.  This can allow off, full forward, or full reverse
+//! control of motors without variable speed.
+//!
+//! # Examples
+//!
+//! Using a bidirectional relay:
+//!
+//! ```
+//! use wpilib::{relay, HalResult};
+//!
+//! # fn main() -> HalResult<()> {
+//! let mut relay = relay::Relay<relay::BothDirections>::new(1)?;
+//!
+//! relay.set(relay::Value::Forward)?;
+//! assert_eq!(relay.get(), relay::Value::Forward);
+//! # }
+//! ```
+
+use std::marker::PhantomData;
+use wpilib_sys::*;
+
+/// Possible values for a bidirectional relay.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub enum Value {
+    /// Both pins set to 0V.
+    Off,
+    /// All pins set to 12V.
+    On,
+    /// Forward pin only set to 12V. Reverse pin set to 0V.
+    Forward,
+    /// Reverse pin only set to 12V. Forward pin set to 0V.
+    Reverse,
+}
+
+/*
+/// Convenience for treating bidirectional and unidirectional relays the same.
+impl From<bool> for Value {
+    fn from(v: bool) -> Self {
+        if v {
+            Value::On
+        } else {
+            Value::Off
+        }
+    }
+}
+*/
+
+/// Marker specifying a Relay goes in both directions.
+pub struct BothDirections;
+/// Marker specifying a Relay goes forward only.
+pub struct ForwardOnly;
+/// Marker specifying a Relay goes reverse only.
+pub struct ReverseOnly;
+
+/// Marker trait for direction type parameter for Relay.
+///
+/// This trait is sealed and cannot be implemented outside of this crate.
+pub trait Direction: private::Sealed {
+    #[doc(hidden)]
+    const FORWARD: bool = true;
+    #[doc(hidden)]
+    const REVERSE: bool = true;
+}
+impl Direction for BothDirections {}
+impl Direction for ForwardOnly {
+    #[doc(hidden)]
+    const REVERSE: bool = false;
+}
+impl Direction for ReverseOnly {
+    #[doc(hidden)]
+    const FORWARD: bool = false;
+}
+
+/// Interface for Spike-style relay outputs.
+///
+/// See the module-level docs for more details.
+#[derive(Debug)]
+pub struct Relay<DIRECTION> {
```

Comment:
Is there a reason `DIRECTION` isn't constrained to implement `Direction`?

Reply:
Not really - probably a sensible thing to do.